ENIGMAsystem/SHELL/Universal_System/Serialization/serialization.h

```diff
@@ -0,0 +1,218 @@
+  } else if constexpr (has_byte_size_method_v<std::decay_t<T>>) {
+    return value.byte_size();
+  } else {
+    return sizeof(T);
+  }
+}
+
+template <typename Base, typename T>
+inline void internal_serialize_primitive_into(std::byte *iter, const T &value) {
+  std::size_t i = sizeof(Base) - 1;
+  std::size_t as_unsigned = utility::bit_cast<Base>(value);
+
+  for (std::size_t j = 0; j < sizeof(Base); j++) {
+    iter[i--] = static_cast<std::byte>(as_unsigned & 0xff);
+    as_unsigned >>= 8;
+  }
+}
+
+template <typename Base, typename T>
+inline std::array<std::byte, sizeof(T)> serialize_primitive(T value) {
+  std::array<std::byte, sizeof(T)> result{};
+  internal_serialize_primitive_into<Base, T>(result.data(), value);
+  return result;
+}
+
+template <typename Base, typename T>
+inline T internal_deserialize_primitive(std::byte *iter) {
+  Base result{};
+  for (std::size_t i = 0; i < sizeof(T); i++) {
+    result = (result << 8) | static_cast<Base>(*(iter + i));
+  }
+  return utility::bit_cast<T>(result);
+}
+
+template <typename T>
+inline void internal_serialize_into(std::byte *iter, T &&value) {
+  if constexpr (_bytes_serialization_has_internal_serialize_into_fn_<std::decay_t<T>>) {
+    enigma::bytes_serialization::internal_serialize_into_fn(iter, value);
+  } else {
+    static_assert(always_false<T>,
+                  "'internal_serialize_into' takes 'variant', 'var', 'std::string', bool, integral, floating types, "
+                  "std::vector, std::map, std::complex, std::set, std::tuple, std::queue, std::stack and std::pair");
+  }
+}
+
+template <typename T>
+inline auto internal_serialize(T &&value) {
+  if constexpr (_bytes_serialization_has_internal_serialize_fn_<std::decay_t<T>>) {
+    return enigma::bytes_serialization::internal_serialize_fn(value);
+  } else if constexpr (has_serialize_method_v<std::decay_t<T>>) {
+    return value.serialize();
+  } else {
+    static_assert(always_false<T>,
+                  "'serialize' takes 'variant', 'var', 'std::string', bool, integral, floating types, std::vector, "
+                  "std::map, std::complex, std::set, std::tuple, std::queue, std::stack and std::pair");
+  }
+}
+
+template <typename T>
+inline T internal_deserialize(std::byte *iter) {
+  if constexpr (_bytes_serialization_has_internal_deserialize_fn_<std::decay_t<T>>) {
+    return enigma::bytes_serialization::internal_deserialize_fn<T>(iter);
+  } else if constexpr (has_deserialize_self_method_v<std::decay_t<T>>) {
+    T result;
+    result.deserialize_self(iter);
+    return result;
+  } else if (has_deserialize_function_v<std::decay_t<T>>) {
+    return internal_deserialize<T>(iter).second;
+  } else {
+    static_assert(always_false<T>,
+                  "'deserialize' takes 'variant', 'var', 'std::string', bool, integral, floating types, std::vector, "
+                  "std::map, std::complex, std::set, std::tuple, std::queue, std::stack and std::pair");
```

Comment:
I think instead of having a message which lists what types have the functions implemented, it may be better to generically say "Type to be deserialized has to implement one of deserialize or deserialize_self!" because the error message cannot catch user types. Serialize would also have a similar message. Same for the JSON functions.